File: web_tool/js/tile_layers.js
var tileLayers = {
    "esri_world_imagery": {
        "location": [[38, -88], 4, "ESRI World Imagery"],
        "tileObject": L.tileLayer('https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}', {
            maxZoom: 20,
            maxNativeZoom: 17,
            attribution: 'Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community'
        }),
        "shapes": null 
    },
    "esri_world_imagery_naip": {
        "location": [[38, -88], 4, "ESRI World Imagery"],
        "tileObject": L.tileLayer('https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}', {
            maxZoom: 20,
            maxNativeZoom: 17,
            attribution: 'Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community'
        }),
        "shapes": null 
    },
    "osm": {
        "location": [[38, -88], 4, "OpenStreetMap"],
        "tileObject": L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
            maxZoom: 17,
            attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
        }),
        "shapes": null 
    },
    "chesapeake": {
        "location": [[38.11437, -75.99980], 10, 'NAIP Imagery - Potential wetlands'],
        "tileObject": L.tileLayer('tiles/chesapeake_test/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 2
        }),
        "shapes": null 
    },
    "demo_set_1": {
        "location": [[39.40625604822793804, -76.5937627969694006], 13, "User study - Demo area"],
        "tileObject": L.tileLayer('tiles/demo_set_1/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": null
    },
    "user_study_1": {
        "location": [[42.406253302897575, -77.12504090737812135], 13, "User study - Area 1"],
        "tileObject": L.tileLayer('tiles/training_set_1/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": null
    },
    "user_study_2": {
        "location": [[42.40625552034823897, -76.87503698687157794], 13, "User study - Area 2"],
        "tileObject": L.tileLayer('tiles/training_set_2/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": null
    },
    "user_study_3": {
        "location": [[42.46875623949721046, -76.50003291357666058], 13, "User study - Area 3"],
        "tileObject": L.tileLayer('tiles/training_set_3/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": null
    },
    "user_study_4": {
        "location": [[43.09375587600650448, -76.18754117285706684], 13, "User study - Area 4"],
        "tileObject": L.tileLayer('tiles/training_set_4/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": null
    },
    "user_study_5": {
        "location": [[42.448269618302362, -75.110429001207137], 13, "User study area"],
        "tileObject": L.tileLayer('tiles/user_study_5/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": [
            {"name": "Area boundary", "shapes_fn": "shapes/user_study_5_outline.geojson", "zone_name_key": null}
        ]
    },
    "philipsburg_mt": {
        "location": [[46.330963, -113.296773], 13, 'Philipsburg, MT'],
        "tileObject": L.tileLayer('tiles/philipsburg/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 20,
            maxNativeZoom: 18,
            minZoom: 13
        }),
        "shapes": null 
    },
    "aceh": { // Planet data from 3-month mosaics covering the Aceh region of Sumatra
        "location": [[3.68745, 97.47070], 7, "Aceh, Sumatra"],
        "tileObject": L.tileLayer('tiles/leuser/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 17,
            maxNativeZoom: 16,
            minZoom: 10
        }),
        "shapes": null 
    },
    "yangon_sentinel": {
        "location": [[16.66177, 96.326427], 10, 'Yangon, Myanmar', "Sentinel Imagery"],
        "tileObject": L.tileLayer('tiles/yangon/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 17,
            maxNativeZoom: 16,
            minZoom: 10
        }),
        "shapes": [
            {"name": "States", "shapes_fn": "shapes/yangon_sentinel_admin_1_clipped.geojson", "zone_name_key": "ST"},
            {"name": "Districts", "shapes_fn": "shapes/yangon_sentinel_admin_2_clipped.geojson", "zone_name_key": "DT"},
            {"name": "Townships", "shapes_fn": "shapes/yangon_sentinel_admin_3_clipped.geojson", "zone_name_key": "TS"},
            {"name": "Wards", "shapes_fn": "shapes/yangon_sentinel_admin_4_clipped.geojson", "zone_name_key": "Ward"}
        ]
    },
    "hcmc_sentinel": {
        "location": [[10.682, 106.752], 11, "Hồ Chí Minh City, Vietnam", "Sentinel Imagery"],
        "tileObject": L.tileLayer('tiles/hcmc_sentinel_tiles/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 17,
            maxNativeZoom: 16,
            minZoom: 10
        }),
        "shapes": [
            {"name": "Admin 1", "shapes_fn": "shapes/hcmc_sentinel_admin_1_clipped.geojson", "zone_name_key": "NAME_1"},
            {"name": "Admin 2", "shapes_fn": "shapes/hcmc_sentinel_admin_2_clipped.geojson", "zone_name_key": "NAME_2"},
            {"name": "Admin 3", "shapes_fn": "shapes/hcmc_sentinel_admin_3_clipped.geojson", "zone_name_key": "NAME_3"}
        ]
    },
    "yangon_lidar": {
        "location": [[16.7870, 96.1450], 15, "Yangon, Myanmar", "LiDAR Imagery"],
        "tileObject": L.tileLayer('tiles/yangon_lidar/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 21,
            maxNativeZoom: 20,
            minZoom: 10
        }),
        "shapes": [
            {"name": "States", "shapes_fn": "shapes/yangon_sentinel_admin_1_clipped.geojson", "zone_name_key": "ST"},
            {"name": "Districts", "shapes_fn": "shapes/yangon_sentinel_admin_2_clipped.geojson", "zone_name_key": "DT"},
            {"name": "Townships", "shapes_fn": "shapes/yangon_sentinel_admin_3_clipped.geojson", "zone_name_key": "TS"},
            {"name": "Wards", "shapes_fn": "shapes/yangon_sentinel_admin_4_clipped.geojson", "zone_name_key": "Ward"}
        ] 
    },
    "hcmc_dg": {
        "location": [[10.838, 106.750], 14, "Thủ Đức District, Hồ Chí Minh City, Vietnam", "Digital Globe Imagery"],
        "tileObject": L.tileLayer('tiles/HCMC/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 21,
            maxNativeZoom: 18,
            minZoom: 14
        }),
        "shapes": [
            {"name": "Admin 1", "shapes_fn": "shapes/hcmc_digital-globe_admin_1_clipped.geojson", "zone_name_key": "NAME_1"},
            {"name": "Admin 2", "shapes_fn": "shapes/hcmc_digital-globe_admin_2_clipped.geojson", "zone_name_key": "NAME_2"},
            {"name": "Admin 3", "shapes_fn": "shapes/hcmc_digital-globe_admin_3_clipped.geojson", "zone_name_key": "NAME_3"}
        ]
    },
    "airbus": {
        "location": [[36.80, -76.12], 14, "Somewhere, VA", "Airbus Imagery"],
        "tileObject": L.tileLayer('tiles/airbus/{z}/{x}/{y}.png', {
            attribution: 'Georeferenced Image', 
            tms:true,
            maxZoom: 21,
            maxNativeZoom: 18,
            minZoom: 13,
            bounds: L.latLngBounds(L.latLng(36.882932, -76.2623637), L.latLng(36.7298842, -76.0249016))
        }),
        "shapes": [
            {"name": "Grid", "shapes_fn": "shapes/airbus-data-grid-epsg4326.geojson", "zone_name_key": null}
        ]
    }
};

var interestingLocations = [
    L.marker([47.60, -122.15]).bindPopup('Bellevue, WA'),
    L.marker([39.74, -104.99]).bindPopup('Denver, CO'),
    L.marker([37.53,  -77.44]).bindPopup('Richmond, VA'),
    L.marker([39.74, -104.99]).bindPopup('Denver, CO'),
    L.marker([37.53,  -77.44]).bindPopup('Richmond, VA'),
    L.marker([33.746526, -84.387522]).bindPopup('Atlanta, GA'),
    L.marker([32.774250, -96.796122]).bindPopup('Dallas, TX'),
    L.marker([40.106675, -88.236409]).bindPopup('Champaign, IL'),
    L.marker([38.679485, -75.874667]).bindPopup('Dorchester County, MD'),
    L.marker([34.020618, -118.464412]).bindPopup('Santa Monica, CA'),
    L.marker([37.748517, -122.429771]).bindPopup('San Fransisco, CA'),
    L.marker([38.601951, -98.329227]).bindPopup('Ellsworth County, KS')
];

File: web_tool/TileLayers.py
import os

import fiona
import shapely.geometry
from enum import Enum

from web_tool.frontend_server import ROOT_DIR

def load_geojson_as_list(fn):
    shapes = []
    crs = None
    with fiona.open(fn) as f:
        crs = f.crs
        for row in f:
            shape = shapely.geometry.shape(row["geometry"])
            shapes.append(shape)
    return shapes, crs

class DataLayerTypes(Enum):
    ESRI_WORLD_IMAGERY = 1
    USA_NAIP_LIST = 2
    CUSTOM = 3

'''
This dictionary defines how the backend tool will return data to the frontend.

An entry is formated like below:

"LAYER NAME": {
    "data_layer_type": DataLayerTypes.ESRI_WORLD_IMAGERY,
    "shapes_fn": None,
    "data_fn": None,
    "shapes": None,  # NOTE: this is always `None` and populated automatically when this file loads (see code at bottom of file)
    "shapes_crs": None  # NOTE: this is always `None` and populated automatically when this file loads (see code at bottom of file)
    "padding": None # NOTE: this is optional and only used in DataLayerTypes.CUSTOM
}

LAYER_NAME - should correspond to an entry in js/tile_layers.js
data_layer_type -  should be an item from the DataLayerTypes enum and describes where the data comes from.
  - If ESRI_WORLD_IMAGERY then the backend will lookup imagery from the ESRI World Imagery basemap and not respond to requests for downloading
  - If USA_NAIP_LIST then the backend will lookup imagery from the full USA tile_index (i.e. how we usually do it) and requests for downloading will be executed on the same tiles
  - If CUSTOM then the backend will query the "shapes_fn" and "data_fn" files for how/what to download, downloading will happen similarlly
shapes_fn - should be a path, relative to `frontend_server.ROOT_DIR`, of a geojson defining shapes over which the data_fn file is valid. When a "download" happens the raster specified by "data_fn" will be masked with one of these shapes.
data_fn - should be a path, relative to `frontend_server.ROOT_DIR`, of a raster file defining the imagery to use
shapes - list of `shapely.geometry.shape` objects created from the shapes in "shapes_fn".
shapes_crs - the CRS of the shapes_fn
padding - NOTE: Optional, only used in DataLayerTypes.CUSTOM - defines the padding used in raster extraction, used to get the required 240x240 input
'''
DATA_LAYERS = {
    "esri_world_imagery": { 
        "data_layer_type": DataLayerTypes.ESRI_WORLD_IMAGERY,
        "shapes": None,
        "data_fn": None,
    },
    "esri_world_imagery_naip": { 
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "osm": {
        "data_layer_type": DataLayerTypes.ESRI_WORLD_IMAGERY,
        "shapes": None,
        "data_fn": None,
    },
    "chesapeake": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "demo_set_1": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "user_study_1": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "user_study_2": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "user_study_3": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "user_study_4": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "user_study_5": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": [
            {"name": "Area boundary", "shapes_fn": "shapes/user_study_5_outline.geojson", "zone_name_key": None}
        ],
        "data_fn": "tiles/user_study_5.tif",
        "padding": 20
    },
    "philipsburg_mt": {
        "data_layer_type": DataLayerTypes.USA_NAIP_LIST,
        "shapes": None,
        "data_fn": None,
    },
    "aceh": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": None,
        "data_fn": None,
    },
    "yangon_sentinel": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": [
            {"name": "Admin 1", "shapes_fn": "shapes/yangon_sentinel_admin_1_clipped.geojson", "zone_name_key": "NAME_1"},
            {"name": "Admin 2", "shapes_fn": "shapes/yangon_sentinel_admin_2_clipped.geojson", "zone_name_key": "NAME_2"},
            {"name": "Admin 3", "shapes_fn": "shapes/yangon_sentinel_admin_3_clipped.geojson", "zone_name_key": "NAME_3"}
        ],
        "data_fn": "tiles/yangon.tif",
        "padding": 1100
    },
    "hcmc_sentinel": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": [
            {"name": "Admin 1", "shapes_fn": "shapes/hcmc_sentinel_admin_1_clipped.geojson", "zone_name_key": "NAME_1"},
            {"name": "Admin 2", "shapes_fn": "shapes/hcmc_sentinel_admin_2_clipped.geojson", "zone_name_key": "NAME_2"},
            {"name": "Admin 3", "shapes_fn": "shapes/hcmc_sentinel_admin_3_clipped.geojson", "zone_name_key": "NAME_3"}
        ],
        "data_fn": "tiles/hcmc_sentinel.tif",
        "padding": 1100
    },
    "yangon_lidar": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": [
            {"name": "Admin 1", "shapes_fn": "shapes/yangon_lidar_admin_1_clipped.geojson", "zone_name_key": "NAME_1"},
            {"name": "Admin 2", "shapes_fn": "shapes/yangon_lidar_admin_2_clipped.geojson", "zone_name_key": "NAME_2"},
            {"name": "Admin 3", "shapes_fn": "shapes/yangon_lidar_admin_3_clipped.geojson", "zone_name_key": "NAME_3"},
            {"name": "Admin 4", "shapes_fn": "shapes/yangon_lidar_admin_4_clipped.geojson", "zone_name_key": "Ward"}
        ],
        "data_fn": "tiles/yangon_lidar.tif",
    },
    "hcmc_dg": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": [
            {"name": "Admin 1", "shapes_fn": "shapes/hcmc_digital-globe_admin_1_clipped.geojson", "zone_name_key": "NAME_1"},
            {"name": "Admin 2", "shapes_fn": "shapes/hcmc_digital-globe_admin_2_clipped.geojson", "zone_name_key": "NAME_2"},
            {"name": "Admin 3", "shapes_fn": "shapes/hcmc_digital-globe_admin_3_clipped.geojson", "zone_name_key": "NAME_3"}
        ],
        "data_fn": "tiles/HCMC.tif",
        "padding": 0
    },
    "airbus": {
        "data_layer_type": DataLayerTypes.CUSTOM,
        "shapes": [
            {"name": "Grid", "shapes_fn": "shapes/airbus-data-grid-epsg4326.geojson", "zone_name_key": None}
        ],
        "data_fn": "tiles/airbus_epsg4326.tif",
        "padding": 0.003
    }
}

for k in DATA_LAYERS.keys():
    if DATA_LAYERS[k]["shapes"] is not None:
        print("Loading shapes for the %s dataset" % (k))
        for zone_layer in DATA_LAYERS[k]["shapes"]:
            fn = os.path.join(ROOT_DIR, zone_layer["shapes_fn"])
            if os.path.exists(fn):
                shapes, crs = load_geojson_as_list(fn)
                zone_layer["shapes_geoms"] = shapes
                zone_layer["shapes_crs"] = crs["init"]
            else:
                print("WARNING: %s doesn't exist, this server will not be able to serve the '%s' dataset" % (fn, k))
